src/ILCompiler.Compiler/src/Compiler/Compilation.cs

```diff
@@ -184,9 +184,20 @@ public void AddCompilationRoot(MethodDesc method, string reason, string exportNa
             public void AddCompilationRoot(TypeDesc type, string reason)
             {
                 if (type.IsGenericDefinition)
+                {
                     _graph.AddRoot(_factory.NecessaryTypeSymbol(type), reason);
+                }
                 else
+                {
                     _graph.AddRoot(_factory.ConstructedTypeSymbol(type), reason);
+
+                    // If the type has a thread static field then we should eagerly create a helper
+                    // to access such fields at runtime. This is required for multi-module compilation.
+                    if (type.IsDefType && (((DefType)type).ThreadStaticFieldSize > 0))
```

Comment:
Yep - when we root a type in library mode, we'll need to carefully root associated statics sections since they're not shareable.